### ISSUE TITLE
Strange synchronous behavior

### ISSUE THREAD

**Issue by mark-anders** [logkit/logkit#10]
I'm looking to switch from CocoaLumberjack to LogKit, with a big reason being able to easily make logging synchronous for debugging.  So far, I like LogKit.

When stepping in the debugger, I am seeing the log output synchronously.  However, when running, I'm seeing very odd results. The test is to output each LogKit type followed by each CocoLumberjack type.  

The result is that sometimes all LogKit messages appear _before_ the CocoaLumberjack ones, sometimes after, and sometimes intermixed.  What's really weird is that the CocoLumberjack statements are async and after the LogKit ones, so how they could ever appear in the order I'm seeing is very odd.  Here's my code:

``` swift
let log = LXLogger(endpoints: [

    LXConsoleEndpoint(
        synchronous: true,
        dateFormatter: LXDateFormatter.timeOnlyFormatter(),
        entryFormatter: LXEntryFormatter({ entry in
            return "\(entry.dateTime)[\(entry.fileName):\(entry.lineNumber)] " +
                    "\(entry.level.uppercaseString)::\(entry.message)"
        })
    ),
])

func testLogging() {
    log.debug("LogKit debug")
    log.info("LogKit info")
    log.notice("LogKit notice")
    log.warning("LogKit warning")
    log.error("LogKit error")
    log.critical("LogKit critical")
    DDLogDebug("CocoaLumberjack debug")
    DDLogInfo("CocoaLumberjack info")
    DDLogVerbose("CocoaLumberjack notice")
    DDLogWarn("CocoaLumberjack warning")
    DDLogError("CocoaLumberjack error")
}
```

and here are some example outputs:

```
08:00:45.215 [AppDelegate.swift:55] DEBUG::CocoaLumberjack debug
08:00:45.215 [AppDelegate.swift:56] INFO::CocoaLumberjack info
08:00:45.215 [AppDelegate.swift:57] VERBOSE::CocoaLumberjack notice
08:00:45.215 [AppDelegate.swift:58] WARNING::CocoaLumberjack warning
08:00:45.215 [AppDelegate.swift:59] ERROR::CocoaLumberjack error
15:00:45.206[AppDelegate.swift:49] DEBUG::LogKit debug
15:00:45.214[AppDelegate.swift:50] INFO::LogKit info
15:00:45.214[AppDelegate.swift:51] NOTICE::LogKit notice
15:00:45.214[AppDelegate.swift:52] WARNING::LogKit warning
15:00:45.214[AppDelegate.swift:53] ERROR::LogKit error
15:00:45.215[AppDelegate.swift:54] CRITICAL::LogKit critical
---------------------------------------------------
08:13:28.951 [AppDelegate.swift:55] DEBUG::CocoaLumberjack debug
15:13:28.935[AppDelegate.swift:49] DEBUG::LogKit debug
15:13:28.950[AppDelegate.swift:50] INFO::LogKit info
15:13:28.950[AppDelegate.swift:51] NOTICE::LogKit notice
15:13:28.951[AppDelegate.swift:52] WARNING::LogKit warning
15:13:28.951[AppDelegate.swift:53] ERROR::LogKit error
15:13:28.951[AppDelegate.swift:54] CRITICAL::LogKit critical
08:13:28.951 [AppDelegate.swift:56] INFO::CocoaLumberjack info
08:13:28.951 [AppDelegate.swift:57] VERBOSE::CocoaLumberjack notice
08:13:28.951 [AppDelegate.swift:58] WARNING::CocoaLumberjack warning
08:13:28.951 [AppDelegate.swift:59] ERROR::CocoaLumberjack error

```

**Comment by justinpawela** [comment]
Thanks for reporting this! It does sound very unexpected.

I am having trouble reproducing your scenario. I created a new (Single View iOS) project, and added both the LogKit and CocoaLumberjack projects. I modeled my code off of yours, and the entirety of it resides in `AppDelegate.swift`. The code is below. Every time I run my tiny test app, I get the log messages in the order I would expect them.

I don't know if my test app is just too simple to reproduce your results. Can you share more info about the project you are using LogKit in? Specifically, are you testing LogKit in a sample application, or trying it out in an larger real-life project? Is there any multithreading or dispatch queues involved? Anything else relevant that might help reproduce this scenario?

Could you also tell me about how you have CocoaLumberjack configured? I'm not super-knowledgeable about CocoaLumberjack, and I don't know if you can configure its `DDTTYLogger` instance in any meaningful way. In my test code below, I just used the `.sharedInstance()`.

Anything about my code that you think might have caused us to have different results?

> `AppDelegate.swift`

``` swift
import UIKit
import LogKit
import CocoaLumberjackSwift


let log = LXLogger(endpoints: [
    LXConsoleEndpoint(
        synchronous: true,
        dateFormatter: LXDateFormatter.timeOnlyFormatter(),
        entryFormatter: LXEntryFormatter({ entry in
            return "\(entry.dateTime)[\(entry.fileName):\(entry.lineNumber)] " +
            "\(entry.level.uppercaseString)::\(entry.message)"
        })
    ),
])


@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {
    var window: UIWindow?

    func application(application: UIApplication, didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {
        DDLog.addLogger(DDTTYLogger.sharedInstance())
        self.testLogging()
        return true
    }

    func testLogging() {
        log.debug("LogKit debug")
        log.info("LogKit info")
        log.notice("LogKit notice")
        log.warning("LogKit warning")
        log.error("LogKit error")
        log.critical("LogKit critical")
        DDLogDebug("CocoaLumberjack debug")
        DDLogInfo("CocoaLumberjack info")
        DDLogVerbose("CocoaLumberjack notice")
        DDLogWarn("CocoaLumberjack warning")
        DDLogError("CocoaLumberjack error")
    }

}
```

My console always outputs in this exact order:

``` text
05:04:16.659[AppDelegate.swift:37] DEBUG::LogKit debug
05:04:16.665[AppDelegate.swift:38] INFO::LogKit info
05:04:16.665[AppDelegate.swift:39] NOTICE::LogKit notice
05:04:16.665[AppDelegate.swift:40] WARNING::LogKit warning
05:04:16.665[AppDelegate.swift:41] ERROR::LogKit error
05:04:16.665[AppDelegate.swift:42] CRITICAL::LogKit critical
2015-12-10 23:04:16:665 LogKitSyncTest[38578:60050498] CocoaLumberjack debug
2015-12-10 23:04:16:665 LogKitSyncTest[38578:60050498] CocoaLumberjack info
2015-12-10 23:04:16:665 LogKitSyncTest[38578:60050498] CocoaLumberjack notice
2015-12-10 23:04:16:665 LogKitSyncTest[38578:60050498] CocoaLumberjack warning
2015-12-10 23:04:16:666 LogKitSyncTest[38578:60050498] CocoaLumberjack error
```

**Comment by mark-anders** [comment]
Hi @justinpawela , well I think I figured out where the problem might be happening.  I put together a sample project under XCode and like you got the correct output.  I then realized that I do a lot of my development using AppCode, so I tried it there.  Sure enough, I ran into the issue. 

So it seems like the issue might be how AppCode handles the output.  However, it is still odd and I'm wondering if there is some issue with how LogKit is outputting the messages. 

I checked in my complete project here:

https://github.com/mark-anders/AppCodeOrderingBug

So you can check it out.  You can also download a 30 day AppCode trial to see if you can reproduce it.

I'll file an AppCode bug to see what they can figure out.

**Comment by mark-anders** [comment]
Hi @justinpawela , a quick followup.  So I traced through what's happening a bit and think that I may have a clue.  I seems like LogKit LXConsoleEndpoint is outputting to stdout. I believe that CocoaLumberjack is outputting to stderr.

So maybe what is happening is that AppCode is prioritizing stderr over stdout.

Does that make sense?

Is there a way to tell the LXConsoleEndpoint to use stderr instead of stdout?

Might be useful.

**Comment by justinpawela** [comment]
@mark-anders Thanks for looking into this more and for posting your findings. I just cloned a copy of your sample project so that I can investigate.

I also dug through the CocoaLumberjack code and confirmed that it is indeed outputting to `stderr`. Then I did some reading on Wikipedia and learned two things:
1. `stderr` is probably more appropriate than `stdout` for logging output
2. As you suspected, this `stdout` vs `stderr` is likely the cause of the problem. From [Wikipedia](https://en.wikipedia.org/wiki/Standard_streams#Standard_error_.28stderr.29):
   
   > Messages appear in the same order as the program writes them, unless buffering is involved. (For example, a common situation is when the standard error stream is unbuffered but the standard output stream is line-buffered; in this case, text written to standard error later may appear on the terminal earlier, if the standard output stream's buffer is not yet full.)
   
   Perhaps AppCode buffers `stdout` where Xcode does not. Regardless, due to point 1, changing LogKit's behavior is probably the most sensible thing to do.

Thankfully, changing to `stderr` is a two-line fix. I have a bug fix update that I'm planning to push sometime today anyway, so this can be included as well.

I am not sure if there is any point in allowing `LXConsoleEndpoint` users the option of choosing `stdout` over (what will be) the default `stderr`. I welcome your thoughts on this.

**Comment by mark-anders** [comment]
Hi @justinpawela , I do not think that it is important to have the option.  I was only suggesting that because I do think having logging to stderr is important and wasn't sure if you were set on having it log to stdout.

Thanks for your promptness in looking into this and look forward to the fix!

**Comment by justinpawela** [comment]
@mark-anders Thanks for your feedback, and all the diagnostic work.

I am working on the bug fix update right now. I am running behind, and may not get to it until tomorrow, but the fix is forthcoming!

**Comment by mark-anders** [comment]
@justinpawela - 2.1.1 indeed fixed the issue.  Thank you.  Oh, and one thing I did not know about AppCode is that stdout is white text while stderr is green.  That should have been my first clue since all of the CocoaLumberjack log statements where green while the LogKit (and Swift print) statements were white.  I had thought that CL was doing some color coding of the log entires rather than it being indicative of the stream.  

tl;dr - all is now green.

Oh, and I'm going to file a feature request for XCodeColor support since it would be nice to have some control over the colors! :-)

**Comment by justinpawela** [comment]
By all means, file it! I have a few potential features in mind that I think would be cool, that I should probably document as well.
